### DISCUSSION TITLE
Add Joystick feature to ESP3D on NodeMCU ESP8266

### DISCUSSION THREAD

**Discussion by Surrizu** [luc-github/ESP3D#705] [Ideas]
Hello Everyone.

I would like to suggest to implement in the ESP3D code a new feature to make possible to connect a joystick shield directlly to the ESP8266 pins.

Joystick Shield expansion board for Arduino:
https://www.amazon.com/Magic-shell-Joystick-Expansion-Keyboard/dp/B07QQXWBCF

With that functionallity, one could move the axis independently by the moviments on the joystick and also one bottom could be assigned to homing command (G28) and another one to bed leveling command (G29).

Regards,
Surrizu.

**Comment by luc-github** [comment]
How many pins this joystick and buttons use ?
Do you have a wiring sample to esp8266 board ?

**Comment by Surrizu** [comment]
Hello.

The Joystick Shield has the following pins (The pins refers to Arduino UNO):

Potentiometers (analog varing from 0 to 5V)
Rx ---> connected to A0
Ry -->  connected to A1

Buttons (digital)
K --> connected to D8
F --> connected to D7
E --> connected to D6
D --> connected to D5
C --> connected to D4
B --> connected to D3
A --> connected to D2

My first idea would be connecting this shield directlly to NodeMCU ESP8266, however NodeMCU has a sole analog input, then it cannont read both Rx and Ry.

So, the following approuch would be using an Arduino UNO to read the pins states and convert this to serial data, sourcing the pins states to NodeMCU.

Here is a proposal:

![Joystick_shiled_UNO_NodeMCU](https://user-images.githubusercontent.com/77760272/148386849-bf555002-8401-4ae7-a9fe-0fd901347087.jpg)

**Reply by luc-github** [reply]
In case of using uno board connected to esp3D board by serial then it is a separate project to do FW for UNO board, what is needed in ESP3D FW is a new Serial as input, do I understand well ?
In case of using ESP board with ESP3D FW, if not enough analog pins with esp8266  then better use esp32  ?

**Reply by Surrizu** [reply]
Yes, a serial interface as input on ESP3D would solve my issue. On Arduíno UNO I can assing each button of the joystick to some GCode and it is necessy that ESP3D just forward the message from UNO to the 3D printer. Like a Hub.

**Reply by akirave** [reply]
Good evening friend, I am working on a similar idea because I wanted to integrate some buttons for additional control for a CNC but I thought I would also put a JOYSTICK and I got a job they did for MARLIN and I see that they are also on the same idea, https://forum .v1engineering.com/t/marlin-joystick-with-g-code-injector/15805................

**Reply by akirave** [reply]
https://proyectosarduinoivan.blogspot.com/2018/01/control-de-cnc-mediante-tarjeta-sd-y.html?m=1
